### PR TITLE
Enbale Int8 path

### PR DESCRIPTION
**Pull request by XiaobingSuper**
- [x] Convolution.

- [x] Relu.

- [x] Pooling.

- [x] BatchNorm.

- [x] Linear.

- [ ]  Sum.

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
How to use?

```
import torch

import intel_pytorch_extension as ipex
import _torch_ipex as core

class MyModel(torch.nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.conv = torch.nn.Conv2d(2, 3, 2, bias=False)
        self.bn = torch.nn.BatchNorm2d(2)
        #self.conv1 = torch.nn.Conv3d(2, 2, 2)
        #self.relu = torch.nn.ReLU(inplace=True)

    def forward(self, x):
        x = self.conv(x)
        #print(x)
        #x = self.bn(x)
        #x = self.conv1(x)
        #x = self.relu(x)
        return x

ipex.enable_auto_mix_precision(mixed_dtype = torch.uint8)
#ipex.enable_auto_optimization(mixed_dtype = torch.int8)

input = torch.randn(2, 2, 4, 4).to(ipex.DEVICE)
model = MyModel().to(ipex.DEVICE)

# calibration
ipex.enable_calibration()
output = model(input)

print(output)
# one iteration is end
core.calibration_reset()
#core.print_observer()

# compute scales
core.add_indictors()
#core.print_indictor()

core.disable_int8_calibration()

output = model(input)
print(output)
output = model(input)
print(output)

```

**Comment by jgong5** [comment]
I suppose we would support serialization and deserialization of the quantization parameters in the future. To align the quant params before each inference, we may consider to use a convenient wrapper around the model. Something like below:
```
class QModel(...):
  def __init__(self, model, q_params):
    self.model = model
    self.q_params = q_params
  def forward(self, *input):
    with self.q_params:
      self.model(*input)
```

**Comment by hongzhen1** [comment]
can we also add one global API with context manager while keep these low-level APIs ?

**Comment by hongzhen1** [comment]
could we also add steps to release the buffer for  observers_ and indicators_ after tensor conversion finishing?

**Comment by XiaobingSuper** [comment]
> can we also add one global API with context manager while keep these low-level APIs ?

done.

**Comment by XiaobingSuper** [comment]
> could we also add steps to release the buffer for observers_ and indicators_ after tensor conversion finishing?

done.

**Comment by XiaobingSuper** [comment]
@hongzhen1 @EikanWang @jgong5 , please help review it.

File: torch_ipex/csrc/auto_opt_config.h [hunk 1]
```diff
@@ -25,15 +29,134 @@ class AutoOptConfig {
...
+      }
+        // zero_points not used now, zero_points = 0 for u8 and 128 for s8.
+        //zero_point = 128;
+      Indicator new_indicator(observers_[i].Id, observers_[i].Name, scales, uint8_used, true);
+      indicators_.push_back(new_indicator);
+    }
+    observers_.clear();
+  }
+
+  inline std::tuple<std::vector<float>, bool> get_indicator_scales(std::vector<bool> uint8_used) {
```
Comment (jgong5) [inline torch_ipex/csrc/auto_opt_config.h:100]: How do we decide what methods are defined in the header file and what are in the source file? Having too many big functions in the header files doesn't sound a good practice, e.g. would slow down the incremental build etc.

File: intel_pytorch_extension_py/__init__.py [hunk 2]
```diff
@@ -31,11 +31,80 @@ def get_auto_optimization():
...
+
+class int8_validate(_DecoratorContextManager):
+    def __init__(self, file_name):
+        self.configure_file = file_name
+
+    def __enter__(self):
+        if not core.get_mix_int8_fp32():
+            raise ValueError("please first run enable_auto_mix_precision(torch.int8) before int8 validation")
+        core.disable_int8_calibration()
+        core.load_indicators_file(self.configure_file)
```
Comment (jgong5) [inline intel_pytorch_extension_py/__init__.py:106]: How about loading the quantization configuration from a memory object instead of a file? We can define separate API for serializing and deserializing the memory object to/from a file. This can separate the concerns of loading/saving the configuration and serialization of it.

File: torch_ipex/csrc/init_python_bindings.cpp [hunk 3]
```diff
@@ -125,7 +126,22 @@ void InitIpexModuleBindings(py::module m) {
...
+
+  m.def("enable_mix_int8_fp32", []() { AutoOptConfig::singleton().set_mix_int8_fp32(true); });
+  m.def("disable_mix_int8_fp32", []() { AutoOptConfig::singleton().set_mix_int8_fp32(false); });
+  m.def("get_mix_int8_fp32", []() { return AutoOptConfig::singleton().get_mix_int8_fp32(); });
+  m.def("enable_int8_calibration", []() { AutoOptConfig::singleton().set_int8_calibration(true); });
+  m.def("disable_int8_calibration", []() { AutoOptConfig::singleton().set_int8_calibration(false); });
+  m.def("get_int8_calibration", []() { return AutoOptConfig::singleton().get_int8_calibration(); });
+  m.def("calibration_reset", []() { AutoOptConfig::singleton().calibration_reset(); });
+  m.def("calibration_reset", []() { AutoOptConfig::singleton().calibration_reset(); });
+  m.def("add_indicators", []() { AutoOptConfig::singleton().add_indicators(); });
```
Comment (jgong5) [inline torch_ipex/csrc/init_python_bindings.cpp:140]: How about "apply_quant_config" instead of "add_indicator"?

File: torch_ipex/csrc/quantization/Observer.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+namespace int8 {
+
+struct Observer {
+  int64_t Id;
+  std::string Name;
+  // the max_values of input and output for one op
+  std::vector<float> max_values;
+};
+
+class Indicator {
```
[the remark this answers is not in the document]
Comment (jgong5) [reply torch_ipex/csrc/quantization/Observer.h:15]: When an OpQuantConfig is defined to an op, is it for some input or output? What would happen if there are multiple inputs and outputs?

File: intel_pytorch_extension_py/__init__.py [hunk 5]
```diff
@@ -31,11 +31,80 @@ def get_auto_optimization():
...
+            raise ValueError("please first run enable_auto_mix_precision(torch.int8) before int8 calibration")
+        core.enable_int8_calibration()
+
+    def __exit__(self, *args):
+        core.disable_int8_calibration()
+        core.add_indicators()
+        core.save_indicators_file(self.configure_file)
+        return False
+
+class int8_validate(_DecoratorContextManager):
```
Comment (jgong5) [inline intel_pytorch_extension_py/__init__.py:98]: I think `int8_evaluate` is a better name here. It aligns with PyTorch's `eval` API. The name "validation" could mean a QA procedure and not accurate.

File: torch_ipex/csrc/cpu/dil/dil/operators/batchnorm.hpp [hunk 6]
```diff
@@ -42,43 +46,94 @@ struct batch_normalization_forward_inference
...
+    #else
+    # pragma omp parallel for schedule(static)
+    #endif
+    #endif
+        for (int c = 0; c < channel_count; c++) {
+          float scale_temp = scale_ptr[c] / std::sqrt(var_ptr[c] + epsilon);
+          scale_shift_buf[c] = scale_temp * dst_scales[0] / src_scales[0];
+          scale_shift_buf[c + channel_count] = (shift_ptr[c] - mean_ptr[c] * scale_temp) * dst_scales[0];
+          expected_mean_ptr[c] = 0.0f;
+          expected_var_ptr[c] = 1.0f;
```
Comment (jgong5) [inline torch_ipex/csrc/cpu/dil/dil/operators/batchnorm.hpp:109]: We may also go a step further to directly compute int8 output from int8 input with fp32 an intermediate data type.

Suppose fp32 BN inference is: `y = ax + b`. And `x = sx(x_i - zx)`, `y = sy(y_i - zy)`, where `x_i` and `y_i` are uint8 asymmetric quantization for `x` and `y`, `sx` and `zx` are scale and zero point for `x_i` while `sy` and `zy` are for `y_i`.

Then we would have: `y_i = uint8_cast[(a*sx/sy) * fp32_cast(x_i) + b + zy - a*sx*zx/sy]`.
